Release v0.6.0 (#1292)
kinvolk/lokomotive: CHANGELOG.md

## v0.6.0 - 2020-12-31

We're happy to announce Lokomotive v0.6.0 (Flying Scotsman).

This release includes several new features, many component updates, and a new platform - [Tinkerbell](https://tinkerbell.org/).

### Changes in v0.6.0

- Remove deprecated CoreOS mentions from AWS ([#1245](https://github.com/kinvolk/lokomotive/pull/1245)) and bare metal ([#1246](https://github.com/kinvolk/lokomotive/pull/1246)).

#### Kubernetes updates

- Update Kubernetes to v1.19.4 and AKS to v1.18.10 ([#1189](https://github.com/kinvolk/lokomotive/pull/1189)).

#### Component updates

> - Update External DNS to v0.7.4 ([#1115](https://github.com/kinvolk/lokomotive/pull/1115)).

Reviewer: Shouldn't the component names be as shown in `lokoctl components list` 
example:  `external-dns`

- Update Prometheus Operator to v0.43.2 ([#1162](https://github.com/kinvolk/lokomotive/pull/1162)).
- Update Calico to v3.17.0 ([#1251](https://github.com/kinvolk/lokomotive/pull/1251)).
- Update aws-ebs-csi-driver to v0.7.0 ([#1135](https://github.com/kinvolk/lokomotive/pull/1135)).

#### Terraform provider updates

- Update Terraform providers to their latest versions ([#1133](https://github.com/kinvolk/lokomotive/pull/1133)).

#### New platforms

- Add support for Tinkerbell platform ([#392](https://github.com/kinvolk/lokomotive/pull/392)).

#### Bug fixes

- Add new worker pools when TLS bootstrap is enabled without remaining stuck ([#1181](https://github.com/kinvolk/lokomotive/pull/1181)).
- Contour: Apply node affinity and tolerations to cert generator job also ([#1161](https://github.com/kinvolk/lokomotive/pull/1161)).
- Improve hardware reservations validation rules on Equinix Metal ([#1186](https://github.com/kinvolk/lokomotive/pull/1186)).
- Equinix Metal: Ensure Calico CRDs get created before use ([#1177](https://github.com/kinvolk/lokomotive/pull/1177)). They were occasionally failing.

> - openebs-operator: Set webhook policy to Ignore ([#1172](https://github.com/kinvolk/lokomotive/pull/1172)). This works around a bug in Kubernetes, see this [comment](https://github.com/openebs/openebs/issues/3046#issuecomment-642558701) for details.

Reviewer: Inconsistency

Either have it `OpenEBS Operator` or `openebs-operator` and similiar for all components.

other example: Line 41
`Contour: ` or `contour:`

#### Features

- Add Packet CCM to Packet platform ([#1155](https://github.com/kinvolk/lokomotive/pull/1155)).
- Contour: Parameterize Envoy scraping interval ([#1229](https://github.com/kinvolk/lokomotive/pull/1229)).
- Expose `--conntrack-max-per-core` kube-proxy flag ([#1187](https://github.com/kinvolk/lokomotive/pull/1187)).
- Add `require_volume_annotation` for restic plugin ([#1132](https://github.com/kinvolk/lokomotive/pull/1132)).
- Print bootkube journal if cluster bootstrap fails ([#1166](https://github.com/kinvolk/lokomotive/pull/1166)). This makes cluster bootstrap problems easier to debug.
- aws-ebs-csi-driver: Add dynamic provisioning, resizing and snapshot options ([#1277](https://github.com/kinvolk/lokomotive/pull/1277)). Now the user has the ability to control the AWS EBS driver to enable or disable provisioning, resizing and snapshotting.

#### Security enhancements

- calico-host-protection: Add custom locked down PSP configuration ([#1274](https://github.com/kinvolk/lokomotive/pull/1274)).

#### Documentation

- Add OpenEBS update guide ([#1163](https://github.com/kinvolk/lokomotive/pull/1163)).

> - Add Rook Ceph update guide ([#1165](https://github.com/kinvolk/lokomotive/pull/1165)),

Reviewer: ```suggestion
- Add Rook Ceph update guide ([#1165](https://github.com/kinvolk/lokomotive/pull/1165)).
```

  If you are running Lokomotive on Equinix Metal (formerly Packet), then install Packet CCM. Export your Packet cluster's project ID and API Key.

  ```bash
  export PACKET_API_KEY=""
  export PACKET_PROJECT_ID=""

  echo "apiKey: $PACKET_API_KEY
projectID: $PACKET_PROJECT_ID" > /tmp/ccm-values.yaml

  helm install packet-ccm --namespace kube-system --values=/tmp/ccm-values.yaml ./lokomotive-0.6.0/assets/charts/control-plane/packet-ccm/
  ```

3. Update all nodes

  This script shipped with release tarball will update all the nodes to run the latest kubelet, latest etcd and add permanent MetalLB labels.

  ```bash
  ./lokomotive-0.6.0/scripts/update/0.5-0.6/update.sh
  ```

4. Export assets directory

  ```sh
  export ASSETS_DIR="assets"
  ```

5. Remove BGP sessions from Terraform state

  If you are running Lokomotive on Equinix Metal (formerly Packet), then run the following commands:

  ```bash
  cd $ASSETS_DIR/terraform
  export WORKER_POOL=""
  export WORKER_COUNT=""

  for ((i=0;i<$WORKER_COUNT;i++)); do
    terraform state rm "module.worker-$WORKER_POOL.packet_bgp_session.bgp[$i]"
  done
  cd -
  ```

  Repeat above steps for each worker pool which has BGP enabled.

> 6. Reload the new config for self-hosted kubelet

Reviewer: ```suggestion
6. Reload the new config for self-hosted kubelet:
```

Author: I removed all the colons https://github.com/kinvolk/lokomotive/pull/1292#discussion_r550126317

Reviewer 2: If we don't use colons, should we put dots at the end of the sentences then?

  ```bash
  kubectl -n kube-system rollout restart ds kubelet
  ```

  If the above command failed, probably you are not running self-hosted kubelet, which is fine because the script in step 3 already updated the host kubelet.

> 7. Remove old asset files

Reviewer: ```suggestion
7. Remove old asset files:
```

  ```sh
  rm -rf $ASSETS_DIR/cluster-assets
  rm -rf $ASSETS_DIR/terraform-modules
  ```

> 8. Update control plane

Reviewer: ```suggestion
8. Update control plane:
```

  ```sh
  lokoctl cluster apply --skip-components -v
  ```

  > **NOTE:** If the upgrade process gets interrupted, rerun `lokoctl cluster apply --skip-components -v`.

  The update process typically takes about 10 minutes.
  After the update, running `lokoctl health` should result in an output similar to the following:

  ```sh
  Node                     Ready    Reason          Message

  lokomotive-controller-0  True     KubeletReady    kubelet is posting ready status
  lokomotive-1-worker-0    True     KubeletReady    kubelet is posting ready status
  lokomotive-1-worker-1    True     KubeletReady    kubelet is posting ready status
  lokomotive-1-worker-2    True     KubeletReady    kubelet is posting ready status
  Name      Status    Message              Error

  etcd-0    True      {"health":"true"}
  ```

#### Updating Contour

> Manually update the CRD before updating the component contour:

Reviewer: ```suggestion
Manually update the CRDs before updating the component contour:
```

- Update `contour` to v1.10.0 ([#1170](https://github.com/kinvolk/lokomotive/pull/1170)).
- Update `experimental-linkerd` to stable-2.9.0 ([#1123](https://github.com/kinvolk/lokomotive/pull/1123)).
- Update `web-ui` to v0.1.3 ([#1237](https://github.com/kinvolk/lokomotive/pull/1237)).
- Update `prometheus-operator` to v0.43.2 ([#1162](https://github.com/kinvolk/lokomotive/pull/1162)).
- Update Calico to v3.17.0 ([#1251](https://github.com/kinvolk/lokomotive/pull/1251)).
- Update `aws-ebs-csi-driver` to v0.7.0 ([#1135](https://github.com/kinvolk/lokomotive/pull/1135)).

#### Terraform provider updates

- Update Terraform providers to their latest versions ([#1133](https://github.com/kinvolk/lokomotive/pull/1133)).

#### New platforms

- Add support for Tinkerbell platform ([#392](https://github.com/kinvolk/lokomotive/pull/392)).

#### Bug fixes

- Add new worker pools when TLS bootstrap is enabled without remaining stuck in the installation phase ([#1181](https://github.com/kinvolk/lokomotive/pull/1181)).
- `contour`: Consistently apply node affinity and tolerations to all scheduled workloads ([#1161](https://github.com/kinvolk/lokomotive/pull/1161)).

> - Equinix Metal: Ensure Calico CRDs get created before use ([#1177](https://github.com/kinvolk/lokomotive/pull/1177)). They were occasionally failing.

Reviewer 2: This doesn't need to be there. v0.5.0 does not include `--wait` yet.

Author: What do you mean v0.5.0 does not include wait? Do you mean it is implemented partially and the user may not see the full benefits just yet?

Reviewer 2: No. I mean users using v0.5.0 are not affected by occasionally failing Calico CRs, because --wait was added after v0.5.0 release.

If buggy --wait would be added in v0.5.0, then this fix should be included in the changelog.

Author: You mean that --wait was added 0.5.0 and this issue with calico failing was fixed immediately. So we don't need to mention it? I don't know if regressions fixed before release should be part of the release notes or not!

Reviewer 2: >You mean that --wait was added 0.5.0 and this issue with calico failing was fixed immediately. So we don't need to mention it? 

Yes, Issue with --wait is not part of v0.5.0 tag/version, so I suggest we remove this from release notes.

>I don't know if regressions fixed before release should be part of the release notes or not!

IMO it shouldn't. This is why you have git history and releases. To reduce amount of history (release notes) for the user to go trough and to have a chance to fix bugs before users notice :)

> - `openebs-operator`: Set webhook policy to Ignore ([#1172](https://github.com/kinvolk/lokomotive/pull/1172)). This works around a bug in Kubernetes, see this [comment](https://github.com/openebs/openebs/issues/3046#issuecomment-642558701) for details.

Reviewer 2: Same here, v0.5.0 does not include e15aec86c7b0c26b8a7135ae4bffd782dbb4c58e.

- `aws-ebs-csi-driver`: Add dynamic provisioning, resizing and snapshot options ([#1277](https://github.com/kinvolk/lokomotive/pull/1277)). Now the user has the ability to control the AWS EBS driver to enable or disable provisioning, resizing and snapshotting.

#### Security enhancements

- `calico-host-protection`: Add custom locked down PSP configuration ([#1274](https://github.com/kinvolk/lokomotive/pull/1274)).

#### Documentation

- Add `openebs-operator` update guide ([#1163](https://github.com/kinvolk/lokomotive/pull/1163)).
- Add `rook-ceph` update guide ([#1165](https://github.com/kinvolk/lokomotive/pull/1165)).

#### Miscellaneous

- Pull control plane images from Quay to avoid hitting Docker Hub pulling limits ([#1226](https://github.com/kinvolk/lokomotive/pull/1226)).
- Bootkube now waits for all control plane charts to converge before exiting, which should make the bootstrapping process more stable ([#1085](https://github.com/kinvolk/lokomotive/pull/1085)).
- Remove deprecated CoreOS mentions from AWS ([#1245](https://github.com/kinvolk/lokomotive/pull/1245)) and bare metal ([#1246](https://github.com/kinvolk/lokomotive/pull/1246)).

> - Improve hardware reservations validation rules on Equinix Metal ([#1186](https://github.com/kinvolk/lokomotive/pull/1186)).

Reviewer 2: Missing newline.

##### Bare-metal

The `cluster.os_channel` parameter got simplified by removing the `flatcar-` prefix.

###### Old

```hcl
os_channel = "flatcar-stable"
```

###### New

```hcl
os_channel = "stable"
```

##### Velero

Velero requires an explicit `provider` field to select the provider.
Example:

```hcl
component `velero` {
  provider = "openebs"

  openebs {
    ...
  }
}
```

#### Cluster update steps

> **NOTE**: Updating multiple Lokomotive versions at a time is not supported. If your cluster is running a version older than `v0.5.0`, update to `v0.5.0` first and only then proceed with the update to `v0.6.0`.

Please perform the following manual steps in your cluster configuration directory.

1. Download the release bundle.

  ```bash
  curl -LO https://github.com/kinvolk/lokomotive/archive/v0.6.0.tar.gz
  tar -xvzf v0.6.0.tar.gz
  ```

2. Install the Packet CCM.

  If you are running Lokomotive on Equinix Metal (formerly Packet), then install Packet CCM. Export your Packet cluster's project ID and API Key.

  ```bash
  export PACKET_API_KEY=""
  export PACKET_PROJECT_ID=""

  echo "apiKey: $PACKET_API_KEY
projectID: $PACKET_PROJECT_ID" > /tmp/ccm-values.yaml

>   helm install packet-ccm --namespace kube-system --values=/tmp/ccm-values.yaml ./lokomotive-0.6.0/assets/charts/control-plane/packet-ccm/

Reviewer 3: I noticed the following error message from the CCM

```
E0111 15:08:09.567379       1 node_lifecycle_controller.go:154] error checking if node test-iago-pool-1-worker-0 is shutdown: providerID cannot be empty string
```

I guess the CCM couldn't populate the providerID in the node object because it's immutable. If we wanted to have it we'd need to recreate the node object.

I'm not sure if this causes problems, can you have a look?

Author: I could not reproduce that error @iaguis `providerID cannot be empty string`.

Reviewer 2: >I guess the CCM couldn't populate the providerID in the node object because it's immutable. If we wanted to have it we'd need to recreate the node object.

CCM does not populate providerID IIRC. It must be done by kubelet. And it will be done when you restart kubelet service (restarting should de-register the node).

Reviewer 3: > I could not reproduce that error @iaguis `providerID cannot be empty string`.

FTR this was updating from v0.5.0. I guess this is what you were trying but wanted to mention it anyway.

> CCM does not populate providerID IIRC. It must be done by kubelet. And it will be done when you restart kubelet service (restarting should de-register the node).

Hm, I wonder how this happened for me then...

Author: > FTR this was updating from v0.5.0. I guess this is what you were trying but wanted to mention it anyway.

Yep I did test it by upgrading from v0.5.0.

  ```

3. Update all nodes.

>   This script shipped with release tarball will update all the nodes to run the latest kubelet, latest etcd and add permanent MetalLB labels.

Reviewer 3: Since this updates to the latest kubelet for the non self-hosted case I wonder if we should flip the `--upgrade-kubelets` flag to being `true` by default, it seems weird that we update kubelets but we don't update the self-hosted kubelet...

Author: It can fail with issue like this: https://github.com/kinvolk/lokomotive/issues/1313. So I don't know if we should set that flag by default.

Reviewer 3: Fair. We should just fix #1313 though but we can do that after the release. Should we mention something in the update instructions so users pass `--upgrade-kubelets` if they are using self hosted kubelet?

- Update `rook` to v1.4.6 ([#1117](https://github.com/kinvolk/lokomotive/pull/1117)).
- Update `velero` to v1.5.2 ([#1131](https://github.com/kinvolk/lokomotive/pull/1131)).
- Update `openebs-operator` to v2.2.0 ([#1095](https://github.com/kinvolk/lokomotive/pull/1095)).
- Update `contour` to v1.10.0 ([#1170](https://github.com/kinvolk/lokomotive/pull/1170)).
- Update `experimental-linkerd` to stable-2.9.0 ([#1123](https://github.com/kinvolk/lokomotive/pull/1123)).
- Update `web-ui` to v0.1.3 ([#1237](https://github.com/kinvolk/lokomotive/pull/1237)).
- Update `prometheus-operator` to v0.43.2 ([#1162](https://github.com/kinvolk/lokomotive/pull/1162)).
- Update Calico to v3.17.0 ([#1251](https://github.com/kinvolk/lokomotive/pull/1251)).
- Update `aws-ebs-csi-driver` to v0.7.0 ([#1135](https://github.com/kinvolk/lokomotive/pull/1135)).

#### Terraform provider updates

- Update Terraform providers to their latest versions ([#1133](https://github.com/kinvolk/lokomotive/pull/1133)).

#### New platforms

- Add support for Tinkerbell platform ([#392](https://github.com/kinvolk/lokomotive/pull/392)).

#### Bug fixes

> - Add new worker pools when TLS bootstrap is enabled without remaining stuck ([#1181](https://github.com/kinvolk/lokomotive/pull/1181)).

Author: yeah sounds incomplete, completed the sentence.

- `contour`: Apply node affinity and tolerations to cert generator job also ([#1161](https://github.com/kinvolk/lokomotive/pull/1161)).

> - Improve hardware reservations validation rules on Equinix Metal ([#1186](https://github.com/kinvolk/lokomotive/pull/1186)).

Author: It closed an issue which was marked as a bug.

> - Improve hardware reservations validation rules on Equinix Metal ([#1186](https://github.com/kinvolk/lokomotive/pull/1186)).

Reviewer 2: Okay, but then the title here needs to be changed to something like: `fixed replacing nodes when reservation IDs changes on Packet platform`. I think we could actually link the same PR both in bug fix section and in enhancement section if you insist keeping it here.

> - Improve hardware reservations validation rules on Equinix Metal ([#1186](https://github.com/kinvolk/lokomotive/pull/1186)).

Author: Moved under `Miscellaneous` section.

1. Download the release bundle

  ```bash
  curl -LO https://github.com/kinvolk/lokomotive/archive/v0.6.0.tar.gz
  tar -xvzf v0.6.0.tar.gz
  ```

2. Install the Packet CCM

  If you are running Lokomotive on Equinix Metal (formerly Packet), then install Packet CCM. Export your Packet cluster's project ID and API Key.

  ```bash
  export PACKET_API_KEY=""
  export PACKET_PROJECT_ID=""

  echo "apiKey: $PACKET_API_KEY
projectID: $PACKET_PROJECT_ID" > /tmp/ccm-values.yaml

  helm install packet-ccm --namespace kube-system --values=/tmp/ccm-values.yaml ./lokomotive-0.6.0/assets/charts/control-plane/packet-ccm/
  ```

3. Update all nodes

  This script shipped with release tarball will update all the nodes to run the latest kubelet, latest etcd and add permanent MetalLB labels.

  ```bash
  ./lokomotive-0.6.0/scripts/update/0.5-0.6/update.sh
  ```

4. Export assets directory

  ```sh
  export ASSETS_DIR="assets"
  ```

5. Remove BGP sessions from Terraform state

  If you are running Lokomotive on Equinix Metal (formerly Packet), then run the following commands:

  ```bash
  cd $ASSETS_DIR/terraform
  export WORKER_POOL=""
  export WORKER_COUNT=""

  for ((i=0;i<$WORKER_COUNT;i++)); do

>     terraform state rm "module.worker-$WORKER_POOL.packet_bgp_session.bgp[$i]"

Reviewer 2: Why not just grep for `packet_bgp_session.bgp` ?

  lokomotive-1-worker-0    True     KubeletReady    kubelet is posting ready status
  lokomotive-1-worker-1    True     KubeletReady    kubelet is posting ready status
  lokomotive-1-worker-2    True     KubeletReady    kubelet is posting ready status
  Name      Status    Message              Error

  etcd-0    True      {"health":"true"}
  ```

#### Update docker log settings

> **NOTE**: This is an optional step if a specific node has humongous-log-generating pods then run these steps on that node.

- SSH into the node and become root with `sudo -s`.
- Create the docker config file:

  ```json
  echo '
  {
    "live-restore": true,
    "log-opts": {
      "max-size": "100m",
      "max-file": "3"
    }
  }
  ' | tee /etc/docker/daemon.json
  ```

- Restart docker daemon:

  > **NOTE**: This will restart all the containers on the node, including kubelet. This step cannot be part of the automatic update script because restarting docker daemon will also kill the update script pod.

  ```bash
  systemctl restart docker
  ```

#### Updating Contour

Manually update the CRDs before updating the component `contour`:

```sh

> kubectl apply -f https://raw.githubusercontent.com/projectcontour/contour/v1.10.0/examples/contour/01-crds.yaml

Reviewer 2: Shouldn't we link to Lokomotive source here?

  kubectl delete pvc data-alertmanager-prometheus-operator-alertmanager-0 -n monitoring
  ```

  You can choose to remove all your existing CRDs (ServiceMonitors, PodMonitors, etc.) if you want to.

3. Remove current `spec.claimRef` values to change the PV's status from Released to Available.

  ```bash
  kubectl patch pv --type json -p='[{"op": "remove", "path": "/spec/claimRef"}]' $(kubectl get pv -o jsonpath='{.items[?(@.spec.claimRef.name=="data-prometheus-prometheus-operator-prometheus-0")].metadata.name}')

  kubectl patch pv --type json -p='[{"op": "remove", "path": "/spec/claimRef"}]' $(kubectl get pv -o jsonpath='{.items[?(@.spec.claimRef.name=="data-alertmanager-prometheus-operator-alertmanager-0")].metadata.name}')
  ```

  > **Note:** To execute the above command, the user must have a cluster wide permission.

4. Make sure that the `prometheus-operator`'s `storage_class` and `prometheus.storage_size` are unchanged during the upgrade process.

5. Proceed to a fresh `prometheus-operator` component installation. The new release should now re-attach your previously released PV with its content.

```
lokoctl component apply prometheus-operator

> ```

Author: @ipochi we suggest that the prometheus operator to be installed only after cluster is updated. But the cluster update process is really long for this release. Can't we suggest to user that they install newer prometheus-operator before update?

Reviewer: Ideally no, because we emphasize that Lokomotive control plane must be updated first and what follows is Lokomotive components.

That being said for instructing to update the component first, it would need to be tested if everything comes up fine, post control plane upgrade.

Reviewer: What other problem do you foresee apart from the cluster update process being lengthy ?

Author: The script that updates the nodes is modifying kubelet, etcd configs and restarting them. This change happens one node at a time hence it is lengthy (the running time can vary depending on the cluster size). Then comes the control plane update part.

Anything going wrong during this period, user has no way of know what hit them or how to mitigate the issue faster because they have no idea at what level things stopped working.

Reviewer: Ok and that would be classified as maintenance window right ? What difference it would make to the process if the Prometheus operator component is updated before or after.

Wouldn't the user have to deal with addtional overhead of an updated prometheus operator if something goes wrong in the update process ?

Author: > Ok and that would be classified as maintenance window right?

Yes I guess that would be counted as maintenance window but there are too many moving parts here.

> What difference it would make to the process if the Prometheus operator component is updated before or after.

Updating component after the cluster update: the release notes from the PR for prometheus operator update process said that certain steps should be performed before cluster update. The new component is installed only after cluster is updated. I think we do that PV patching to make sure that data is not lost due to prometheus-operator component update (especially due to the change of names).

Updating component before cluster update: yes that the question I am interested in, why can't we do all of it before?

> Wouldn't the user have to deal with addtional overhead of an updated prometheus operator if something goes wrong in the update process ?

Here we come into picture. We figure out all the stuff that can go wrong. Provide mitigation plans and if nothing works user can still rollback to older `prometheus-operator` component, unlike the updating cluster where rollback would be much much harder.

Reviewer: I don't have any issues with updating the prometheus operator component beore the cluster update, provided this process is tested and there are no consequences.

Reviewer 2: >The script that updates the nodes is modifying kubelet, etcd configs and restarting them. This change happens one node at a time hence it is lengthy (the running time can vary depending on the cluster size). Then comes the control plane update part.

Controlplane should be upgraded first, then nodes. This is how Kubernetes version skew recommends it.

About the Prometheus Operator, IMO it should be updated after the controlplane, as we don't test new Prometheus version with old Kubernetes. The same applies for all other components.

Author: I am addressing the updating kubelet etcd and kubelet version after control plane has updated. But I still think that prometheus-operator should be updated before the cluster update.

- Bootkube now waits for all control plane charts to converge before exiting, which should make the bootstrapping process more stable ([#1085](https://github.com/kinvolk/lokomotive/pull/1085)).
- Remove deprecated CoreOS mentions from AWS ([#1245](https://github.com/kinvolk/lokomotive/pull/1245)) and bare metal ([#1246](https://github.com/kinvolk/lokomotive/pull/1246)).
- Improve hardware reservations validation rules on Equinix Metal ([#1186](https://github.com/kinvolk/lokomotive/pull/1186)).

### Updating from v0.5.0

#### Configuration syntax changes

##### AWS

Removed the undocumented `cluster.os_name` parameter, since Lokomotive supports Flatcar Container Linux only.

##### Bare-metal

The `cluster.os_channel` parameter got simplified by removing the `flatcar-` prefix.

###### Old

```hcl
os_channel = "flatcar-stable"
```

###### New

```hcl
os_channel = "stable"
```

##### Velero

Velero requires an explicit `provider` field to select the provider.
Example:

```hcl
component `velero` {
  provider = "openebs"

  openebs {
    ...
  }
}
```

#### Updating Prometheus Operator

> Upgrading this component incurs down time and some steps to be executed manually before upgrading:

Reviewer 3: I'm missing some context on why we need to do this. Also, let's be consistent on using update/upgrade.

```suggestion
Due to a change in the upstream Helm chart, updating the Prometheus Operator component incurs down time. We do this before updating the cluster so no visibility is lost while the cluster update is happening.
```

1. Patch the `PersistentVolume` created/used by the `prometheus-operator` component to `Retain` claim policy.

  ```bash
  kubectl patch pv -p '{"spec":{"persistentVolumeReclaimPolicy":"Retain"}}' $(kubectl get pv -o jsonpath='{.items[?(@.spec.claimRef.name=="data-prometheus-prometheus-operator-prometheus-0")].metadata.name}')

  kubectl patch pv -p '{"spec":{"persistentVolumeReclaimPolicy":"Retain"}}' $(kubectl get pv -o jsonpath='{.items[?(@.spec.claimRef.name=="data-alertmanager-prometheus-operator-alertmanager-0")].metadata.name}')
  ```

  > **Note:** To execute the above command, the user must have a cluster wide permission.

2. Uninstall the `prometheus-operator` release and delete the existing `PersistentVolumeClaim`, and verify `PersistentVolume` become `Released`.

  ```bash
  lokoctl component delete prometheus-operator
  ```

  ```bash
  kubectl delete pvc data-prometheus-prometheus-operator-prometheus-0 -n monitoring
  kubectl delete pvc data-alertmanager-prometheus-operator-alertmanager-0 -n monitoring
  ```

>   You can choose to remove all your existing CRDs (ServiceMonitors, PodMonitors, etc.) if you want to.

Reviewer 2: I think we should remove this line, as we discussed.

Reviewer 3: Yes. Also, this is pretty vague: when and why would you want to remove CRDs?

Author: > Yes. Also, this is pretty vague: when and why would you want to remove CRDs?

The user should not remove the CRs but rather update them. Prometheus operator has this convention of labels  it checks like if there is a certain label on configmap then that is queried for grafana dashboard, or if there is a certain label on the ServiceMonitor only then it is scraped etc. Now these labels have changed with the move to new prometheus operator chart so user is advised to udpate the components. Since these components ship the configs with new labels.

3. Remove current `spec.claimRef` values to change the PV's status from Released to Available.

  ```bash
  kubectl patch pv --type json -p='[{"op": "remove", "path": "/spec/claimRef"}]' $(kubectl get pv -o jsonpath='{.items[?(@.spec.claimRef.name=="data-prometheus-prometheus-operator-prometheus-0")].metadata.name}')

  kubectl patch pv --type json -p='[{"op": "remove", "path": "/spec/claimRef"}]' $(kubectl get pv -o jsonpath='{.items[?(@.spec.claimRef.name=="data-alertmanager-prometheus-operator-alertmanager-0")].metadata.name}')
  ```

  > **Note:** To execute the above command, the user must have a cluster wide permission.

4. Make sure that the `prometheus-operator`'s `storage_class` and `prometheus.storage_size` are unchanged during the upgrade process.

5. Proceed to a fresh `prometheus-operator` component installation. The new release should now re-attach your previously released PV with its content.

```
lokoctl component apply prometheus-operator
```

> 6. If monitoring was enabled for `rook`, `contour`, `metallb` components, then update the said components as well after the cluster is updated.

Reviewer 3: ```suggestion
6. If monitoring was enabled for `rook`, `contour`, `metallb` components, make sure you update them as well after the cluster is updated.
```

> **NOTE**: Updating multiple Lokomotive versions at a time is not supported. If your cluster is running a version older than `v0.5.0`, update to `v0.5.0` first and only then proceed with the update to `v0.6.0`.

Please perform the following manual steps in your cluster configuration directory.

1. Download the release bundle.

  ```bash
  curl -LO https://github.com/kinvolk/lokomotive/archive/v0.6.0.tar.gz
  tar -xvzf v0.6.0.tar.gz
  ```

2. Install the Packet CCM.

  If you are running Lokomotive on Equinix Metal (formerly Packet), then install Packet CCM. Export your Packet cluster's project ID and API Key.

  ```bash
  export PACKET_AUTH_TOKEN=""
  export PACKET_PROJECT_ID=""

  echo "apiKey: $PACKET_AUTH_TOKEN
projectID: $PACKET_PROJECT_ID" > /tmp/ccm-values.yaml

  helm install packet-ccm --namespace kube-system --values=/tmp/ccm-values.yaml ./lokomotive-0.6.0/assets/charts/control-plane/packet-ccm/
  ```

3. Update node config.

  This script shipped with release tarball will add permanent MetalLB labels and kubelet config to use CCM.

  > **NOTE**: Please edit this script to disable updating certain nodes. Modify `update_other_nodes` function as required.

  ```bash

>   ./lokomotive-0.6.0/scripts/update/0.5-0.6/update.sh false

Reviewer 3: This false parameter seems pretty magic, how about something like:

```suggestion
  UPDATE_BOOTSTRAP_COMPONENTS=false
  ./lokomotive-0.6.0/scripts/update/0.5-0.6/update.sh $UPDATE_BOOTSTRAP_COMPONENTS
```

  ```

> 4. Reload the new config for self-hosted kubelet.
> 
>   ```bash
>   kubectl -n kube-system rollout restart ds kubelet
>   ```
> 
>   If the above command failed, probably you are not running self-hosted kubelet, which is fine because the script in step 3 already updated the host kubelet.

Reviewer 3: I don't like this "probably" there, how about this?

```suggestion
4. If you're using the self-hosted kubelet, reload its config.

  ```bash
  kubectl -n kube-system rollout restart ds kubelet
  ```

  If you're unsure you can run the command as it's harmless if you're not using the self-hosted kubelet.
```

  lokoctl cluster apply --skip-components -v
  ```

  > **NOTE:** If the update process gets interrupted, rerun above command.

  > **NOTE:** If you are running self-hosted kubelet then append the above command with flag `--upgrade-kubelets`.

  The update process typically takes about 10 minutes.
  After the update, running `lokoctl health` should result in an output similar to the following:

  ```bash
  Node                     Ready    Reason          Message

  lokomotive-controller-0  True     KubeletReady    kubelet is posting ready status
  lokomotive-1-worker-0    True     KubeletReady    kubelet is posting ready status
  lokomotive-1-worker-1    True     KubeletReady    kubelet is posting ready status
  lokomotive-1-worker-2    True     KubeletReady    kubelet is posting ready status
  Name      Status    Message              Error

  etcd-0    True      {"health":"true"}
  ```

9. Update the bootstrap components: kubelet and etcd.

  This script shipped with release tarball will update all the nodes to run the latest kubelet and etcd.

  > **NOTE**: Please edit this script to disable updating certain nodes. Modify `update_other_nodes` function as required.

  ```bash

>   ./lokomotive-0.6.0/scripts/update/0.5-0.6/update.sh true

Reviewer 3: Please use a bash variable like mentioned above.

  ```

> 10. Reload the new config for self-hosted kubelet.

Reviewer 3: Do we need to run this twice? We've run it already in step 4.

Reviewer 3: If we need to run it twice, please change the text as mentioned in the previous comment.

Author: This time the reload is to update the version of kubelet. Earlier one was to update a piece of config.

  ```bash
  kubectl -n kube-system rollout restart ds kubelet
  ```

  If the above command failed, probably you are not running self-hosted kubelet, which is fine because the script in step 3 already updated the host kubelet.

> #### Update docker log settings

Reviewer 3: ```suggestion
#### Update Docker log settings
```

> > **NOTE**: This is an optional step if a specific node has humongous-log-generating pods then run these steps on that node.

Reviewer 3: I don't like this phrasing. I'd rather have something like this. WDYT?

```suggestion
We've added log rotation to the Docker daemon running on cluster nodes. However, this only takes effect in new nodes. For this to apply to existing cluster nodes, you need to manually configure each node.
```

- SSH into the node and become root with `sudo -s`.

> - Create the docker config file:

Reviewer 3: ```suggestion
- Create the Docker config file:
```

  ```json
  echo '
  {
    "live-restore": true,
    "log-opts": {
      "max-size": "100m",
      "max-file": "3"
    }
  }
  ' | tee /etc/docker/daemon.json
  ```

> - Restart docker daemon:

Reviewer 3: ```suggestion
- Restart the Docker daemon:
```

>   > **NOTE**: This will restart all the containers on the node, including kubelet. This step cannot be part of the automatic update script because restarting docker daemon will also kill the update script pod.

Reviewer 3: ```suggestion
  > **NOTE**: This will restart all the containers on the node, including the kubelet. This step cannot be part of the automatic update script because restarting the Docker daemon will also kill the update script pod.
```

  ```bash
  systemctl restart docker
  ```

#### Updating Contour

Manually update the CRDs before updating the component `contour`:

```bash
kubectl apply -f https://raw.githubusercontent.com/kinvolk/lokomotive/v0.6.0/assets/charts/components/contour/crds/01-crds.yaml
```

Update the component:

```bash
lokoctl component apply contour
```

#### Updating Velero

Manually update the CRDs before updating the component `velero`:

```bash

> kubectl apply -f ./lokomotive-0.6.0/assets/charts/components/velero/crds/

Reviewer 3: Why not use the github URL?

If you want to use the local path I'd use `$ASSETS_DIR`.

Author: > Why not use the github URL?

There are many files in that dir to use the github URL.

> If you want to use the local path I'd use $ASSETS_DIR.

Assets directory does not have velero configuration.

> ## v0.6.0 - 2020-12-XX

Reviewer 4: I think this will be 2021 now 🙈

  ```

9. Update the bootstrap components: kubelet and etcd.

  This script shipped with release tarball will update all the nodes to run the latest kubelet and etcd.

  > **NOTE:** Please edit this script to disable updating certain nodes. Modify `update_other_nodes` function as required.

  ```bash
  UPDATE_BOOTSTRAP_COMPONENTS=true
  ./lokomotive-0.6.0/scripts/update/0.5-0.6/update.sh $UPDATE_BOOTSTRAP_COMPONENTS
  ```

10. If you're using the self-hosted kubelet, reload its config.

  > **NOTE:** If you're unsure you can run the command as it's harmless if you're not using the self-hosted kubelet.

  ```bash
  kubectl -n kube-system rollout restart ds kubelet
  ```

#### Update Docker log settings

We've added log rotation to the Docker daemon running on cluster nodes. However, this only takes effect in new nodes. For this to apply to existing cluster nodes, you need to manually configure each node.

- SSH into the node and become root with `sudo -s`.
- Create the Docker config file:

  ```json
  echo '
  {
    "live-restore": true,
    "log-opts": {
      "max-size": "100m",
      "max-file": "3"
    }
  }
  ' | tee /etc/docker/daemon.json
  ```

- Restart the Docker daemon:

>   > **NOTE:** This will restart all the containers on the node, including the kubelet. This step cannot be part of the automatic update script because restarting the Docker daemon will also kill the update script pod.

Reviewer 2: Nit: we should drain the node before doing that to make the process smoother.

- Update Linkerd to stable-2.9.0 ([#1123](https://github.com/kinvolk/lokomotive/pull/1123)).
- Update Web UI to v0.1.3 ([#1237](https://github.com/kinvolk/lokomotive/pull/1237)).
- Update Prometheus Operator to v0.43.2 ([#1162](https://github.com/kinvolk/lokomotive/pull/1162)).
- Update Calico to v3.17.0 ([#1251](https://github.com/kinvolk/lokomotive/pull/1251)).

#### Terraform provider updates

- Update Terraform providers to their latest versions([#1133](https://github.com/kinvolk/lokomotive/pull/1133)).

#### Other updates

- Update aws-ebs-csi-driver to v0.7.0 ([#1135](https://github.com/kinvolk/lokomotive/pull/1135)).

#### New platforms

- Add support for Tinkerbell as a platform ([#392](https://github.com/kinvolk/lokomotive/pull/392)).

#### Bug fixes

- Fix bug adding worker pools with TLS bootstrap enabled ([#1181](https://github.com/kinvolk/lokomotive/pull/1181)).
- contour: apply node affinity and tolerations also to certgen job ([#1161](https://github.com/kinvolk/lokomotive/pull/1161)).
- Improve hardware reservations validation rules on Equinix Metal ([#1186](https://github.com/kinvolk/lokomotive/pull/1186)).

> - Equinix Metal: ensure Calico CRDs gets created before use ([#1177](https://github.com/kinvolk/lokomotive/pull/1177)). They were failing occasionally.

Reviewer 5: ```suggestion
- Equinix Metal: ensure Calico CRDs get created before use ([#1177](https://github.com/kinvolk/lokomotive/pull/1177)). They were failing occasionally.
```

- openebs-operator: set webhook policy to Ignore ([#1172](https://github.com/kinvolk/lokomotive/pull/1172)). This works around a bug in Kubernetes, see this [comment](https://github.com/openebs/openebs/issues/3046#issuecomment-642558701) for details.

> - Don't run control plane components as DaemonSets on single control plane node clusters ([#1193](https://github.com/kinvolk/lokomotive/pull/1193)).

Reviewer 5: I'm not sure this is a bugfix. Is it?

Author: Yes it was bug fix, cause this was a problem only for single node clusters and we were counting it wrong. Even for single node clusters we would bring up at least two replicas of each control plane components.

> - calico-host-protection: Add custom locked down PSP configuration ([#1274](https://github.com/kinvolk/lokomotive/pull/1274)).

Reviewer 5: This looks like a security enhancement to me rather than a bugfix.

#### Features

- Add Packet CCM to Packet platform ([#1155](https://github.com/kinvolk/lokomotive/pull/1155)).
- contour: parameterize Envoy scraping interval ([#1229](https://github.com/kinvolk/lokomotive/pull/1229)).
- Expose --conntrack-max-per-core kube-proxy flag ([#1187](https://github.com/kinvolk/lokomotive/pull/1187)).

> - Add `require_volume_annotation` for restic plugin ([#1132](https://github.com/kinvolk/lokomotive/pull/1132)). This PR is based on top of #1131 and would need to be merged first.

Reviewer 5: ```suggestion
- Add `require_volume_annotation` for restic plugin ([#1132](https://github.com/kinvolk/lokomotive/pull/1132)).
```

I'm not sure the last sentence belongs in the release notes. Is there an open TODO item there? What is the context here?

Author: I think it was a left over from formatting the release script output.

> - Print bootkube journal if cluster bootstrap fails ([#1166](https://github.com/kinvolk/lokomotive/pull/1166)). This makes easier to debug problems with cluster bringup.

Reviewer 4: > This makes easier to debug problems with cluster bringup.
	
This sounds a bit off, what do you think?

Author: How about now?

> - Print bootkube journal if cluster bootstrap fails ([#1166](https://github.com/kinvolk/lokomotive/pull/1166)). This makes easier to debug problems with cluster bringup.

Reviewer 5: ```suggestion
- Print bootkube journal if cluster bootstrap fails ([#1166](https://github.com/kinvolk/lokomotive/pull/1166)). This makes cluster bootstrap problems easier to debug.
```

> - aws-ebs-csi-driver: Add dynamic provisioning, resizing, snapshot options ([#1277](https://github.com/kinvolk/lokomotive/pull/1277)). Now the user has ability to control the AWS EBS driver to enable or disable provisioning, resizing, snapshotting.

Reviewer 4: ```suggestion
- aws-ebs-csi-driver: Add dynamic provisioning, resizing, snapshot options ([#1277](https://github.com/kinvolk/lokomotive/pull/1277)). Now the user has ability to control the AWS EBS CSI driver to enable or disable provisioning, resizing, snapshotting.
```

> - aws-ebs-csi-driver: Add dynamic provisioning, resizing, snapshot options ([#1277](https://github.com/kinvolk/lokomotive/pull/1277)). Now the user has ability to control the AWS EBS driver to enable or disable provisioning, resizing, snapshotting.

Reviewer 5: ```suggestion
- aws-ebs-csi-driver: Add dynamic provisioning, resizing and snapshot options ([#1277](https://github.com/kinvolk/lokomotive/pull/1277)). Now the user has the ability to control the AWS EBS driver to enable or disable provisioning, resizing and snapshotting.
```

> ## Documentation

Reviewer 5: ```suggestion
#### Documentation
```

> - Add OpenEBS upgrade guide ([#1163](https://github.com/kinvolk/lokomotive/pull/1163)).

Reviewer 4: ```suggestion
- Add OpenEBS update guide ([#1163](https://github.com/kinvolk/lokomotive/pull/1163)).
```

> - Add Rook Ceph upgrade guide ([#1165](https://github.com/kinvolk/lokomotive/pull/1165)),

Reviewer 4: ```suggestion
- Add Rook Ceph update guide ([#1165](https://github.com/kinvolk/lokomotive/pull/1165)).
```

> #### Others

Reviewer 5: ```suggestion
#### Miscellaneous
```

> - Pull control plane images from Quay ([#1226](https://github.com/kinvolk/lokomotive/pull/1226)). To avoid hitting Docker Hub pulling limits.

Reviewer 5: ```suggestion
- Pull control plane images from Quay to avoid hitting Docker Hub pulling limits ([#1226](https://github.com/kinvolk/lokomotive/pull/1226)).
```

> - Bootkube now waits for all controlplane charts to converge before exiting, which should make bootstrapping process more stable ([#1085](https://github.com/kinvolk/lokomotive/pull/1085)).

Reviewer 5: ```suggestion
- Bootkube now waits for all control plane charts to converge before exiting, which should make the bootstrapping process more stable ([#1085](https://github.com/kinvolk/lokomotive/pull/1085)).
```

> ### Upgrading from v0.5.0

Reviewer 4: ```suggestion
### Updating from v0.5.0
```

#### Configuration syntax changes

##### AWS

> The undocumented `cluster.os_name` parameter was removed since only Flatcar
> Container Linux is supported.

Reviewer 5: ```suggestion
The undocumented `cluster.os_name` parameter was removed since only Flatcar Container Linux is supported.
```

The line break looks misplaced.

##### Bare-metal

The `cluster.os_channel` parameter got simplified by removing the `flatcar-` prefix.

> ###### Old:

Reviewer 5: ```suggestion
###### Old
```

Either we use colons for both "Old" and "New" or we don't :wink:

```hcl
os_channel = "flatcar-stable"
```

###### New

```hcl
os_channel = "stable"
```

##### Velero

Velero requires an explicit `provider` field to select the provider.
Example:

```hcl
component `velero` {
  provider = "openebs"

  openebs {
    ...
  }
}
```

> #### Cluster upgrade steps

Reviewer 4: ```suggestion
#### Cluster update steps
```

> Updating multiple versions at a time is not supported so, if your cluster is older, upgrade to `v0.5.0` and only then proceed with the upgrade to `v0.6.0`.

Reviewer 5: ```suggestion
Updating multiple Lokomotive versions at a time is not supported. If your cluster is running a version older than `v0.5.0`, upgrade to `v0.5.0` first and only then proceed with the upgrade to `v0.6.0`.
```

Improve phrasing and reduce ambiguity.

> Some manual steps need to be performed when updating. In your cluster configuration directory, follow these steps:

Reviewer 5: ```suggestion
Some manual steps need to be performed when updating. In your cluster configuration directory, follow the steps below.
```

I find "follow these steps" with the colon confusing given that we have just one step (exporting the assets dir var) before the next heading.

> Starting from your cluster directory, export your assets directory name used in your platform configuration. It will be used in next steps:

Reviewer 5: >It will be used in next steps:

I would either move this sentence *after* the code block or omit it altogether.

Author: eliminated, it was redundant I guess.

```sh
export ASSETS_DIR="assets"
```

> ##### Only for Packet/Equinix Metal

Reviewer 5: The phrasing we use in the headings here is inconsistent to me:

- Only for Packet/Equinix Metal
- General upgrade steps
- Upgrading native kubelets and etcd (optional)
- Updating Contour

Can we use a more consistent phrasing? It should be very clear to the users which steps they should perform and which they shouldn't.

> 1. Disable state management for BGP sessions

Reviewer 5: ```suggestion
1. Remove BGP sessions from Terraform state
```

```bash
cd $ASSETS_DIR/terraform
export WORKER_POOL=""
export WORKER_COUNT=""

for ((i=0;i<$WORKER_COUNT;i++)); do
  terraform state rm "module.worker-$WORKER_POOL.packet_bgp_session.bgp[$i]"
done
cd -
```

> Repeat above steps for each worker pool running MetalLB with BGP enabled on it.

Reviewer 5: ```suggestion
Repeat step 1 for each worker pool which has BGP enabled.
```

> 2. Install Packet CCM

Reviewer 5: ```suggestion
2. Install the Packet CCM
```

  Export your Packet cluster's project ID and API Key.

  ```bash
  export PACKET_API_KEY=""
  export PACKET_PROJECT_ID=""

  echo "apiKey: $PACKET_API_KEY
projectID: $PACKET_PROJECT_ID" > /tmp/ccm-values.yaml

>   # TODO: Update this to 0.6.0

Reviewer 5: @surajssd looks like a leftover TODO :warning:

  # curl -LO https://github.com/kinvolk/lokomotive/archive/v0.5.0.tar.gz
  # tar -xvzf v0.5.0.tar.gz
  # helm install packet-ccm --namespace kube-system --values=/tmp/ccm-values.yaml lokomotive-0.5.0/assets/charts/control-plane/packet-ccm/

>   # TODO: Remove this before merging

Reviewer 5: @surajssd looks like a leftover TODO :warning:

  curl -LO https://github.com/kinvolk/lokomotive/archive/master.zip
  unzip master.zip
  helm install packet-ccm --namespace kube-system --values=/tmp/ccm-values.yaml lokomotive-master/assets/charts/control-plane/packet-ccm/
  ```

> 3. Add permanent labels to the nodes running MetalLB, just run the following command and label will be added:

Reviewer 5: ```suggestion
3. Add permanent labels to the nodes running MetalLB
```

- I think we should avoid words such as "just", "simply" etc. They don't add clarity to the docs and are basically implying "this should be easy for you to do" :slightly_smiling_face: 
- We are inconsistent with these numbered headings throughout the release notes. If it's a heading it shouldn't end with a colon. If it isn't a heading it should be a full sentence ending with a period.

> Create the script which will be used to modify the env var file. So even if the node is restarted these labels will persist:

Reviewer 5: - What is "the env var file"? Is it important for me as a user to know this? If so, we should explain, otherwise we might want to remove the explanation and simply tell the users which commands to run.
- Starting a sentence with "so" sounds wrong to me at the English level when the sentence continues elaborates on the previous one.
- "These labels" assumes context. We should either explain which labels we are talking about and why the user needs to worry about them, or alternatively include just a series of commands which the user needs to follow.

Author: Everything is moved to a separate script that will be shipped with code tar ball.

```bash

> echo '

Reviewer 5: I would strongly suggest we commit the script to the repo and let the user run it.

#!/bin/bash

set -euo pipefail

echo -e "Updating Kubelet env file...\nOld Kubelet env file:\n"
cat /etc/kubernetes/kubelet.env

label=$(grep ^NODE_LABELS /etc/kubernetes/kubelet.env)
label_prefix="${label::-1}"
augmented_label="${label_prefix},lokomotive.alpha.kinvolk.io/bgp-enabled=true\""

> # This copy is needed because sed tries to create a new file. And this changes inode and docker does not allow it.

Reviewer 5: "And" after period.

cp /etc/kubernetes/kubelet.env /kubelet.env
sed -i "s|^NODE_LABELS.*|${augmented_label}|g" /kubelet.env
cat /kubelet.env > /etc/kubernetes/kubelet.env

echo -e "\nNew Kubelet env file:\n"
cat /etc/kubernetes/kubelet.env

echo -e "\nUpdating Kubelet config creator file...\nOld Kubelet config creator file:\n"
cat /etc/kubernetes/configure-kubelet-cgroup-driver

cp /etc/kubernetes/configure-kubelet-cgroup-driver /configure-kubelet-cgroup-driver
sed -i "s|^EOF|providerID: external|g" /configure-kubelet-cgroup-driver
echo "EOF" >> /configure-kubelet-cgroup-driver
cat /configure-kubelet-cgroup-driver > /etc/kubernetes/configure-kubelet-cgroup-driver

echo -e "\nNew Kubelet config creator file:\n"
cat /etc/kubernetes/configure-kubelet-cgroup-driver

echo -e "\nRestarting Kubelet...\n"
exec nsenter -a -t 1 /bin/sh -c "systemctl restart kubelet && systemctl status --no-pager kubelet"
' | tee /tmp/run
```

> Store above script in a configmap in a separate namespace:

Reviewer 5: Missing article.

```bash
kubectl create ns update-host-files
kubectl create -n update-host-files cm script --from-file /tmp/run --dry-run=client -o yaml | kubectl apply -f -
```

> Run this command to create pods on each node that is running metallb:

Reviewer 5: - Does the user needs to select which nodes to run the commands on? If not, let's drop the extra explanation and say "Run the following command".
- MetalLB should be capitalized correctly.

```bash
for nodename in $(kubectl get nodes -l metallb.lokomotive.io/my-asn=65000,lokomotive.alpha.kinvolk.io/bgp-enabled!=true -ojsonpath='{.items[*].metadata.name}')

> # for nodename in $(kubectl get nodes -l metallb.lokomotive.io/my-asn=65000 -ojsonpath='{.items[*].metadata.name}')

Reviewer 5: We should either uncomment this line or remove it.

do
  echo "
apiVersion: v1
kind: Pod
metadata:
  labels:
    run: update-host-files-$nodename
  name: update-host-files-$nodename
  namespace: update-host-files
spec:
  containers:
  - image: registry.fedoraproject.org/fedora:32
    name: update-host-files
    imagePullPolicy: IfNotPresent
    securityContext:
      privileged: true
    args:
    - bash
    - /tmp/script/run
    volumeMounts:
    - name: kubelet-env
      mountPath: /etc/kubernetes/kubelet.env
    - name: kubelet-config
      mountPath: /etc/kubernetes/configure-kubelet-cgroup-driver
    - name: script
      mountPath: /tmp/script/
  nodeName: $nodename
  restartPolicy: Never
  hostPID: true
  volumes:
  - name: kubelet-env
    hostPath:
      path: /etc/kubernetes/kubelet.env
  - name: kubelet-config
    hostPath:
      path: /etc/kubernetes/configure-kubelet-cgroup-driver
  - name: script
    configMap:
      name: script
" | kubectl apply -f -

  while true
  do
    complete=$(kubectl -n update-host-files get pod update-host-files-$nodename -o jsonpath='{.status.containerStatuses[-1].state.terminated.exitCode}')

>     #

Reviewer 5: ?

    if [ -z "${complete}" ]; then
      echo "waiting for pod update-host-files-$nodename to finish"
    else
      kubectl -n update-host-files logs update-host-files-$nodename
      break
    fi
    sleep 1
  done

  kubectl label node $nodename lokomotive.alpha.kinvolk.io/bgp-enabled=true
done
```

4. Reload the new config for self-hosted kubelet

>   If you are running self-hosted kubelet then restart it:

Reviewer 5: How do I know if I'm running a self-hosted kubelet? We should either explain or better yet - make the command figure this out on the user's behalf and run conditionally.

  ```bash
  kubectl -n kube-system rollout restart ds kubelet
  ```

> #### General upgrade steps

Reviewer 4: ```suggestion
#### General update steps
```

1. Remove old asset files:

  ```sh
  rm -rf $ASSETS_DIR/cluster-assets
  rm -rf $ASSETS_DIR/terraform-modules
  ```

2. Finally, run the following:

```sh
lokoctl cluster apply --skip-components -v
```

> **NOTE:** If the upgrade process gets interrupted. Re-run `lokoctl cluster apply --skip-components -v` to proceed.

Reviewer 5: ```suggestion
**NOTE:** If the upgrade process gets interrupted, rerun `lokoctl cluster apply --skip-components -v`.
```

The upgrade process typically takes about 10 minutes.
After the upgrade, running `lokoctl health` should result in an output similar to the following:

```sh
Node                     Ready    Reason          Message

lokomotive-controller-0  True     KubeletReady    kubelet is posting ready status
lokomotive-1-worker-0    True     KubeletReady    kubelet is posting ready status
lokomotive-1-worker-1    True     KubeletReady    kubelet is posting ready status
lokomotive-1-worker-2    True     KubeletReady    kubelet is posting ready status
Name      Status    Message              Error

etcd-0    True      {"health":"true"}
```

> #### Upgrading native kubelets and etcd (optional)

Reviewer 4: ```suggestion
#### Updating native kubelets and etcd (optional)
```

> - Manually upgrade etcd following the steps mentioned in the doc

Reviewer 4: ```suggestion
- Manually update etcd following the steps mentioned in the doc
```

>   [here](https://github.com/kinvolk/lokomotive/blob/v0.5.0/docs/how-to-guides/upgrade-etcd.md).

Reviewer 4: ```suggestion
  [here](https://github.com/kinvolk/lokomotive/blob/v0.6.0/docs/how-to-guides/upgrade-etcd.md).
```

>   [here](https://github.com/kinvolk/lokomotive/blob/v0.5.0/docs/how-to-guides/upgrade-etcd.md).

Reviewer 5: ```suggestion
- Manually upgrade etcd following the steps mentioned [here](https://github.com/kinvolk/lokomotive/blob/v0.5.0/docs/how-to-guides/upgrade-etcd.md).
```

> - Manually upgrade the kubelet running on the nodes, by following the steps mentioned in the doc

Reviewer 4: ```suggestion
- Manually update the kubelet running on the nodes, by following the steps mentioned in the doc
```

>   [here](https://github.com/kinvolk/lokomotive/blob/v0.5.0/docs/how-to-guides/upgrade-bootstrap-kubelet.md).

Reviewer 4: ```suggestion
  [here](https://github.com/kinvolk/lokomotive/blob/v0.6.0/docs/how-to-guides/upgrade-bootstrap-kubelet.md).
```

>   [here](https://github.com/kinvolk/lokomotive/blob/v0.5.0/docs/how-to-guides/upgrade-bootstrap-kubelet.md).

Reviewer 5: ```suggestion
- Manually upgrade the kubelet running on the nodes by following the steps mentioned [here](https://github.com/kinvolk/lokomotive/blob/v0.5.0/docs/how-to-guides/upgrade-bootstrap-kubelet.md).
```

#### Updating Contour

Due to some CRD changes, some manual steps are required to upgrade Contour:

```sh
kubectl apply -f https://raw.githubusercontent.com/projectcontour/contour/v1.10.0/examples/contour/01-crds.yaml
```

> Now it is safe to upgrade:

Reviewer 5: I find this kind of tone changes confusing. I understand it's safe to upgrade, but do I *need* to upgrade? In which cases should I run this command?

```sh
lokoctl component apply contour
```

> #### Upgrading OpenEBS

Reviewer 4: ```suggestion
#### Updating OpenEBS
```

Follow the [OpenEBS upgrade guide](https://kinvolk.io/docs/lokomotive/0.6/how-to-guides/upgrade-openebs/).

> #### Upgrading Rook Ceph

Reviewer 4: ```suggestion
#### Updating Rook Ceph
```

Follow the [Rook Ceph upgrade guide](https://kinvolk.io/docs/lokomotive/0.6/how-to-guides/upgrade-rook-ceph/).

> #### Upgrading other components

Reviewer 4: ```suggestion
#### Updating other components
```

Please perform the following manual steps in your cluster configuration directory.

1. Download the release bundle.

  ```bash
  curl -LO https://github.com/kinvolk/lokomotive/archive/v0.6.0.tar.gz
  tar -xvzf v0.6.0.tar.gz
  ```

2. Install the Packet CCM.

  If you are running Lokomotive on Equinix Metal (formerly Packet), then install Packet CCM. Export your Packet cluster's project ID and API Key.

  ```bash
  export PACKET_AUTH_TOKEN=""
  export PACKET_PROJECT_ID=""

  echo "apiKey: $PACKET_AUTH_TOKEN
projectID: $PACKET_PROJECT_ID" > /tmp/ccm-values.yaml

  helm install packet-ccm --namespace kube-system --values=/tmp/ccm-values.yaml ./lokomotive-0.6.0/assets/charts/control-plane/packet-ccm/
  ```

3. Update node config.

  This script shipped with release tarball will add permanent MetalLB labels and kubelet config to use CCM.

  > **NOTE:** Please edit this script to disable updating certain nodes. Modify `update_other_nodes` function as required.

  ```bash
  UPDATE_BOOTSTRAP_COMPONENTS=false
  ./lokomotive-0.6.0/scripts/update/0.5-0.6/update.sh $UPDATE_BOOTSTRAP_COMPONENTS
  ```

> 4. If you're using the self-hosted kubelet, apply the `--cloud-provider` flag to it.

Reviewer 2: There is no need to do that manually I think. `lokoctl cluster apply` should apply that automatically.

Author: This step is being run before the cluster is updated, hence I included this step.

Reviewer 2: But why do we need to run this before cluster update?

Author: Because the self-hosted kubelets and bootstrap kubelets are on the same config.

Reviewer 2: So long you don't remove the Node object, it shouldn't matter. I don't think there is a strict requirement to do that right after patching bootstrap kubelet. It simplifies and speeds up the upgrade process, which is already very long and complicated...

Author: If a user missed to use the `--upgrade-kubelets` with `lokoctl cluster apply` it is highly possible that the subsequent nodes added won't join the cluster, so this steps acutally guards against that mistake.

Reviewer 2: Then we should require using `--upgrade-kubelets`?

Author: Yes, there is a note saying if you are using self-hosted kubelet then use this flag.

Reviewer 2: Then I think it should be fine to remove this step if we _require_ using `--upgrade-kubelets`.

Reviewer 3: Given https://github.com/kinvolk/lokomotive/issues/1313 I think we should not require `--upgrade-kubelets` until that's solved.

But I wonder if we can include this step in the update script checking if the self-hosted kubelet is running.

Author: Yeah with #1313 we will need this manual step.

  > **NOTE:** SSH into a given node in a separate console upfront, before starting to follow steps.

  This step ensures that you don't see any abrupt changes. Any workloads running on this node are evicted and scheduled to other nodes. The node is marked as unschedulable after running this command.

  ```bash
  export nodename=""
  kubectl drain --ignore-daemonsets $nodename
  ```

  Delete the node object:

  ```bash
  kubectl delete node $nodename
  ```

  SSH into the node:

  ```bash
  sudo systemctl restart kubelet
  ```

  Make the node schedulable:

  ```bash
  while ! kubectl uncordon $nodename; do sleep 1; done
  ```

#### Update Docker log settings

We've added log rotation to the Docker daemon running on cluster nodes. However, this only takes effect in new nodes. For this to apply to existing cluster nodes, you need to manually configure each node.

- Drain the node.

  This step ensures that you don't see any abrupt changes. Any workloads running on this node are evicted and scheduled to other nodes. The node is marked as unschedulable after running this command.

  ```bash

>   kubectl drain --ignore-daemonsets <node name>

Reviewer 2: I don't like that we ask user to drain all nodes twice. I think we could merge those instructions somehow.

Author: Yeah, I don't like either. But these are two unrelated tasks and I don't want to confound user by giving so much information at once.

Author: Or make them feel like they need to do all the steps at once or make them mix things up.

Reviewer 2: Maybe at least we can add a cross-reference between tasks, so one can perform both at the same time?

Author: This is moved to a new doc.

}
```

> This release also changes the default `cluster.oidc.client_id` value from `gangway` to `clusterauth`. 

Reviewer 2: This should be committed separately.
